### PR TITLE
Alerting: separate out silence auth service preconditions checks

### PR DESCRIPTION
**Pull request by JacobsonMT**
Will be useful for subsequent PR that adds metadata to silence response

**What is this feature?**

Separates out precondition checks from silence auth service `AuthorizeXSilence` methods. 

**Why do we need this feature?**

This is so precondition checks can be used separately in a subsequent PR that will add permission metadata information to the GET silence responses.

**Special notes for your reviewer:**

This is mostly a  no-op refactor but introduces a new short circuit for silence reads when a user has wildcard folder permissions.

### CODE REVIEW

File: pkg/services/ngalert/accesscontrol/silences.go [hunk 1]
```diff
@@ -97,157 +97,219 @@ func NewSilenceService(ac ac.AccessControl, store RuleUIDToNamespaceStore) *Sile
...
-		return fmt.Errorf("resolve rule UID to folder UID: %w", err)
+// authorizeReadSilence checks if user has access to read a silence given precondition checks have passed.
+func (s SilenceService) authorizeReadSilence(ctx context.Context, user identity.Requester, silence *silenceWithFolder) error {
+	if silence.ruleUID == nil {
+		return nil // No rule metadata means that this is a general silence and at this point the user can read them
 	}
-	if folderUID == "" { // if we did not find folder by rule UID then it does not exist.
-		return NewAuthorizationErrorGeneric(fmt.Sprintf("read silence for rule %s", *ruleUID))
+
+	if silence.folderUID == "" { // if we did not find folder by rule UID then it does not exist.
```
Comment (yuri-tceretian) [inline pkg/services/ngalert/accesscontrol/silences.go:183]: If I read this correctly, in a situation when a rule is assigned to a folder that does not exist (which can happen), nobody will be able to see this silence, even "admin". I think we should return it if user can read all silences (ideally, when wildcard folder scope too)

File: pkg/services/ngalert/accesscontrol/silences.go [hunk 2]
```diff
@@ -97,157 +97,219 @@ func NewSilenceService(ac ac.AccessControl, store RuleUIDToNamespaceStore) *Sile
...
-			continue
-		}
-		hasAccess, ok := namespacesByAccess[ns]
+	for _, silWithFolder := range silencesWithFolders {
+		hasAccess, ok := namespacesByAccess[silWithFolder.folderUID]
 		if !ok {
-			hasAccess, err = s.HasAccess(ctx, user, readRuleSilenceEvaluator(ns))
-			if err != nil {
-				return nil, err
+			hasAccess = s.authorizeReadSilence(ctx, user, silWithFolder) == nil
```
Comment (yuri-tceretian) [inline pkg/services/ngalert/accesscontrol/silences.go:130]: creation of the authz error is a bit more costly than regular error, I wonder if we can have just bool here. This is not a blocker, though.

File: pkg/services/ngalert/accesscontrol/silences.go [hunk 3]
```diff
@@ -97,157 +97,219 @@ func NewSilenceService(ac ac.AccessControl, store RuleUIDToNamespaceStore) *Sile
...
+		return fmt.Errorf("resolve rule UID to folder UID: %w", err)
+	}
+
+	return s.authorizeReadSilence(ctx, user, silWithFolder[0])
+}
+
+// authorizeReadSilencePreConditions checks necessary preconditions for reading silences. Returns true if user can
+// read all silences. Returns error if user does not have access to read any silences.
+func (s SilenceService) authorizeReadSilencePreConditions(ctx context.Context, user identity.Requester) (bool, error) {
+	canAll, err := s.HasAccess(ctx, user, readAllSilencesEvaluator)
```
Comment (yuri-tceretian) [inline pkg/services/ngalert/accesscontrol/silences.go:162]: I wonder if we can hack into user's permissions and check if user has wildcard scope, and set canAll to true if it has.

**Review by yuri-tceretian — APPROVED** [review]
My comments are non-blocking and can be implemented in a follow-up PR that already there. 
Feel free to merge it as is. LGTM

File: pkg/services/ngalert/accesscontrol/silences.go (hunk 3 above)
Comment (JacobsonMT) [reply pkg/services/ngalert/accesscontrol/silences.go:162]: Added silence read folder wildcard scope to the readAllSilencesEvaluator: https://github.com/grafana/grafana/pull/87998/commits/c214702f11b26fe788c4ae415e6ac74e570dfe41

This is not technically checking for all forms of wildcards, but should cover the main one without having to manually check user permissions.

**Pull request merged**